### ISSUE TITLE
XFS extent allocator consuming >10ms on schemas with one large table

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#3167]
Parent issue: #3119 

Due to an XFS heuristic, if all files are in one (or a few) directories, then block allocation can become very slow. This is because XFS divides the disk into a few allocation groups (AGs), and each directory allocates preferentially from a single AG. That AG can become filled long before the disk is full.

**Comment by avikivity** [comment]
Advice from the XFS mailing list (Dave Chinner): use the inode32 mount option, which will break the affinity between a directory and an AG. The installer should set it, and we should also recommend it to users who set up the disks themselves.

**Comment by avikivity** [comment]
Existing users can adjust `/etc/fstabs` and update the running system with

    mount /var/lib/scylla -o remount,inode32

**Comment by syuu1228** [comment]
@avikivity we can create fstab entry with inode32 option on scylla_raid_setup, for newly installed instance.
Maybe we also can refuse starting scylla if startup script did not detect inode32 mount option on data directory, like iotune settings.

**Comment by avikivity** [comment]
We can start by just warning about it, it's just a performance option, and we don't have enough experience with it.

**Comment by avikivity** [comment]
Implementation suggestion (from #3645):

We can work around the problem by placing each sstable in its own directory: instead of

    /var/lib/scylla/data/keyspace1/standard1-96ec6d80795811e8b330000000000001/la-66869-big-Data.db
use

    /var/lib/scylla/data/keyspace1/standard1-96ec6d80795811e8b330000000000001/66869.sstable/la-66869-big-Data.db

so each sstable gets its own directory. As a bonus, moving or deleting a single sstable using command-line tools becomes simple.

We should of course continue to read old-layout sstables and also write using the old layout until the entire cluster recognizes the new format (to allow downgrades).

**Comment by raphaelsc** [comment]
Perhaps, for the sake of simplicity, we could link the sstable stored in its own directory to parent directory, as sealing step of sstable write, such that existing code for loading remains the same and migration is made easier.

**Comment by avikivity** [comment]
It's indeed possible, we can even remove the .sstable directory afterwards.

Although, it's nice to have one entry per sstable in the main data directory instead of 7X.

**Comment by bhalevy** [comment]
See https://github.com/bhalevy/scylla/tree/issues/3167/RFCv1 for possible implementation, with no support for an option to disable the feature.

It's not clear how such an option should be implemented.  We might also want to have a similar option for the MC sstable format.  Should we have a generic option for controlling all storage format-changing features during cluster upgrade so they are enabled in an orchestrated way only once the whole cluster has been upgraded?
